### PR TITLE
Fix pylint errors in tcms/testcases/views

### PR DESCRIPTION
**Pull request by ivo0126**

### CODE REVIEW

File: tcms/testcases/views.py [hunk 1]
```diff
@@ -588,12 +603,14 @@ class SimpleTestCaseView(TemplateView):
...
 
     template_name = 'case/get_details.html'
+    case_id = None
+    review_mode = None
 
     # NOTES: what permission is proper for this request?
-    def get(self, request, case_id):
-        self.case_id = case_id
+    def get(self, request, *args, **kwargs):
+        self.case_id = args[0]
```
Comment (atodorov) [inline tcms/testcases/views.py]: Revert this change!

File: tcms/testcases/views.py [hunk 2]
```diff
@@ -588,12 +603,14 @@ class SimpleTestCaseView(TemplateView):
...
         self.review_mode = request.GET.get('review_mode')
-        return super(SimpleTestCaseView, self).get(request, case_id)
+        return super(SimpleTestCaseView, self).get(request, self.case_id)
```
Comment (atodorov) [inline tcms/testcases/views.py]: Use `super()`

File: tcms/testcases/views.py [hunk 3]
```diff
@@ -617,45 +634,48 @@ def get_context_data(self, **kwargs):
...
     """Display case runs list when expand a plan from case page"""
 
     template_name = 'case/get_case_runs_by_plan.html'
+    case_id = None
+    plan_id = None
 
     # FIXME: what permission here?
-    def get(self, request, case_id):
-        self.case_id = case_id
+    def get(self, request, *args, **kwargs):
```
Comment (atodorov) [inline tcms/testcases/views.py]: Revert this change!

File: tcms/testcases/views.py [hunk 4]
```diff
@@ -630,32 +662,20 @@ def get(self, request, case_id):
         self.plan_id = int(plan_id) if plan_id is not None else None
 
         this_cls = TestCaseCaseRunListPaneView
-        return super(this_cls, self).get(request, case_id)
+        return super(this_cls, self).get(request, self.case_id)
```
Comment (atodorov) [inline tcms/testcases/views.py]: Use `super()`

Comment (atodorov) [reply tcms/testcases/views.py]: Does pylint complain if we use case_id as parameter instead of self.case_id ? If there's no difference I prefer a smaller commit and not including lines which don't fix any issues or not change behavior.

File: tcms/testcases/views.py [hunk 5]
```diff
@@ -58,74 +58,78 @@ def plan_from_request_or_none(request, pk_enough=False):
...
 
 def group_case_bugs(bugs):
     """Group bugs using bug_id."""
     bugs = itertools.groupby(bugs, lambda b: b.bug_id)
-    bugs = [(pk, list(_bugs)) for pk, _bugs in bugs]
+    bugs = []
+
+    for key, _bugs in bugs:
+        bugs.append((key, list(_bugs)))
+
```
Comment (atodorov) [inline tcms/testcases/views.py]: This is breaking existing behavior, so you will have to add a test for that, and also fix it. 

The problem lies in how list comprehension is evaluated. First the right part is evaluated, where `bugs` is the input parameter, which is an iterator (coming from line 106). The data is constructed into a list and assigned again to the `bugs` variable, changing its data type.

In your implementation L106 `bugs` is an iterator. Then L107 makes `bugs` an empty list. Then L109 executes a loop over the empty list with name `bugs`, which doesn't go into any iterations so on L112 you remove an empty result when the input parameter wasn't empty.

Comment (atodorov) [reply tcms/testcases/views.py]: Note: the entire queryset to iterator (with group by that can be done via QS) to a list of tuples, containing list of bugs is beyond me. Let's fix the obvious issue first (with a test case) and later work on refining this function.

Note2: Move this change into another PR and ping me to get some help about writing a test for this. Once that is merged we can return here and rebase.

File: tcms/testcases/views.py [hunk 6]
```diff
@@ -771,33 +800,39 @@ def get(request, case_id):
...
         'assignee', 'case',
         'case', 'case_run_status').order_by('run__plan')
     runs_ordered_by_plan = itertools.groupby(tcrs, lambda t: t.run.plan)
     # FIXME: Just don't know why Django template does not evaluate a generator,
     # and had to evaluate the groupby generator manually like below.
-    runs_ordered_by_plan = [(k, list(v)) for k, v in runs_ordered_by_plan]
-    case_run_plans = [k for k, v in runs_ordered_by_plan]
+    runs_ordered_by_plan = []
+    for key, value in runs_ordered_by_plan:
+        runs_ordered_by_plan.append((key, list(value)))
```
Comment (atodorov) [inline tcms/testcases/views.py]: This is the same error as seen above when working with the bugs. L828 essentially makes the list of runs empty and the loops below will never iterate!

This breaks the UI, TestCase get view, Case Runs tab is now empty! You will have to create a test for that (in a 2nd PR once the 1st one is done) and fix it. 

Note: ^^^ you have to work in smaller chunks next time and learn to also manually test and inspect the application when you introduce changes.

File: tcms/testcases/views.py (hunk 5 above)
Comment (ivo0126) [reply tcms/testcases/views.py]: a variable called bugs is causing a bug lol

File: tcms/testcases/views.py [hunk 7]
```diff
@@ -220,17 +220,17 @@ def _continue(self):
...
 
                     return HttpResponseRedirect(
-                        reverse('testcases-edit', args=[tc.case_id]))
+                        reverse('testcases-edit', args=[test_case.case_id]))
 
                 def _addanother(self):
                     form = NewCaseForm(initial=default_form_parameters)
 
-                    if tp:
+                    if test_plan:
```
Comment (atodorov) [inline tcms/testcases/views.py]: This is `self.plan`. Don't use the global variable `test_plan`.

File: tcms/testcases/views.py [hunk 8]
```diff
@@ -220,17 +220,17 @@ def _continue(self):
...
                         form.populate(product_id=self.plan.product_id)
 
                     return form
 
-                def _returntocase(self):
+                def returntocase(self):
```
Comment (atodorov) [inline tcms/testcases/views.py]: Why do you rename only this mathod and not the rest ?

File: tcms/testcases/views.py [hunk 9]
```diff
@@ -249,31 +249,30 @@ def _returntoplan(self):
                                                    args=[self.plan.pk]))
 
             # Genrate the instance of actions
```
Comment (atodorov) [inline tcms/testcases/views.py]: fix the typo in comment

File: tcms/testcases/views.py [hunk 10]
```diff
@@ -457,14 +461,15 @@ def get_selected_testcases(request):
...
         plan = plan_from_request_or_none(request)
         cases, _search_form = query_testcases_from_request(request, plan)
         return cases
-    else:
-        pks = [int(pk) for pk in REQ.getlist('case')]
-        return TestCase.objects.filter(pk__in=pks)
+    keys = []
+    for key in method.getlist('case'):
+        keys.append(int(key))
+    return TestCase.objects.filter(pk__in=keys)
```
Comment (atodorov) [inline tcms/testcases/views.py]: If `method.getlist('case')` returns a list of strings that represent primary keys then `.filter(pk__in=...)` should work just fine with it. Slit this change into a separete pull request, create a test for it and try to remove the intermediate for loop.
